71_RayTracingPipeline/app_resources/raytrace.rchit.hlsl

```diff
@@ -0,0 +1,157 @@
+                n1 = normalize(nbl::hlsl::spirv::unpackSnorm4x8(v1).xyz);
+                n2 = normalize(nbl::hlsl::spirv::unpackSnorm4x8(v2).xyz);
+            }
+            break;
+        case OT_SPHERE:
+        case OT_CYLINDER:
+        case OT_ARROW:
+        case OT_CONE:
+        {
+                uint32_t v0 = vk::RawBufferLoad < uint32_t > (normalVertexBufferAddress + i0 * vertexStride);
+                uint32_t v1 = vk::RawBufferLoad < uint32_t > (normalVertexBufferAddress + i1 * vertexStride);
+                uint32_t v2 = vk::RawBufferLoad < uint32_t > (normalVertexBufferAddress + i2 * vertexStride);
+
+                n0 = normalize(unpackNormals3x10(v0));
+                n1 = normalize(unpackNormals3x10(v1));
+                n2 = normalize(unpackNormals3x10(v2));
+            }
+            break;
+        case OT_RECTANGLE:
+        case OT_DISK:
+        case OT_ICOSPHERE:
+        default:
+        {
+                n0 = normalize(vk::RawBufferLoad <
+                float3 > (normalVertexBufferAddress + i0 * vertexStride));
+                n1 = normalize(vk::RawBufferLoad <
+                float3 > (normalVertexBufferAddress + i1 * vertexStride));
+                n2 = normalize(vk::RawBufferLoad <
+                float3 > (normalVertexBufferAddress + i2 * vertexStride));
+            }
+    }
+
+    float3 barycentrics = float3(0.0, bary);
+    barycentrics.x = 1.0 - barycentrics.y - barycentrics.z;
+
+    VertexData data;
+    data.position = barycentrics.x * p0 + barycentrics.y * p1 + barycentrics.z * p2;
+    data.normal = normalize(barycentrics.x * n0 + barycentrics.y * n1 + barycentrics.z * n2);
+    return data;
+}
+
+[shader("closesthit")]
+void main(inout ColorPayload p, in BuiltInTriangleIntersectionAttributes attribs)
+{
+    const int instID = InstanceID();
+    const int primID = PrimitiveIndex();
+    const STriangleGeomInfo geom = vk::RawBufferLoad < STriangleGeomInfo > (pc.triangleGeomInfoBuffer + instID * sizeof(STriangleGeomInfo));
+    const VertexData vertexData = fetchVertexData(instID, primID, geom, attribs.barycentrics);
+    const float32_t3 worldPosition = mul(ObjectToWorld3x4(), float32_t4(vertexData.position, 1));
+    const float32_t3 worldNormal = normalize(mul(vertexData.normal, WorldToObject3x4()).xyz);
+
+    RayLight cLight;
+    cLight.inHitPosition = worldPosition;
+    CallShader(pc.light.type, cLight);
+
+    float32_t3 diffuse = computeDiffuse(geom.material, cLight.outLightDir, worldNormal);
+    float32_t3 specular = float32_t3(0, 0, 0);
+    float32_t attenuation = 1;
+
+    if (dot(worldNormal, cLight.outLightDir) > 0)
+    {
+        RayDesc rayDesc;
+        rayDesc.Origin = WorldRayOrigin() + WorldRayDirection() * RayTCurrent();
+        rayDesc.Direction = cLight.outLightDir;
+        rayDesc.TMin = 0.01;
+        rayDesc.TMax = cLight.outLightDistance;
+
+        uint flags = RAY_FLAG_SKIP_CLOSEST_HIT_SHADER;
```

Comment:
what's the reason for not having `RAY_FLAG_ACCEPT_FIRST_HIT_AND_END_SEARCH` here?